test_psa_constant_names: support key agreement, better code structure

gilles-peskine-arm opened the pull request:
This pull request consists mostly of changes to `test_psa_constant_names.py`:

* Many small refactorings of `test_psa_constant_names.py`. I initially did them in preparation for reusing major parts of the code in another script. This other script is not stable yet and I'm not sure if I need it after all. Since the refactorings are at worst harmless and at best useful to make the code easier to maintain, I'm submitting them.
* Fix the collection of curves, which I discovered to be broken.
* Error out if a test case uses a macro name which isn't defined in a header covered by this script.
* Add support for key agreement, which was long overdue.
* In `test_suite_psa_crypto_metadata`, add a few more test cases for KDF.

Patater commented on tests/scripts/test_psa_constant_names.py:
```diff
@@ -348,13 +359,9 @@ def run_tests(options, inputs):
-        c, e = do_test(options, inputs, type_word, names)
+    for type_word in ['status', 'algorithm', 'ecc_curve', 'dh_group',
+                      'key_type', 'key_usage']:
+        c, e = do_test(options, inputs, type_word)
```
In commit "Move the type_word->name_set mapping into its own method", which says it has no change on behavior, we don't need to call `get_names()` anywhere? For example, while passing a names parameter to `do_test()`

Patater commented on tests/scripts/test_psa_constant_names.py:
```diff
@@ -324,7 +324,18 @@ def normalize(expr):
+    values = run_c(options, type_word, expressions)
+    return expressions, values
+
+def do_test(options, inputs, type_word):
```
In commit "Move value collection into its own function", which says it makes no change to behavior, don't we need to update all callers of `do_test()`? We are changing the parameters `do_test()` makes in a commit that doesn't change the callers, so I don't see how this wouldn't change behavior.

gilles-peskine-arm commented:
I rewrote the history in two places:

* I fixed the split between "Move the type_word->name_set mapping into its own method" and "Move value collection into its own function" to make each commit correct as pointed out in the review.
* I modified the commit "add_test_case_line: data-driven dispatch" significantly to better separate data and code and to avoid reaching pylint's threshold for too many branches when the kdf/ka cases are added.

Patater commented on tests/scripts/test_psa_constant_names.py:
```diff
@@ -200,26 +229,35 @@ def parse_header(self, filename):
+                # Keep the expression as one to test as an algorithm.
+                function = 'other_algorithm'
+        if function in self.table_by_test_function:
+            sets.append(self.table_by_test_function[function])
```
Do we want to call `accept_test_case_line()` if `function` was not in `self.table_by_test_function`? I see `accept_test_case_line()` throws an exception instead of returning `False` for not accepted test cases.

gilles-peskine-arm replied:
Good question. If we add a function to `test_suite_psa_crypto_metadata`, what do we want to happen here? I don't think the current behavior is correct: a new function whose name does not end in `_algorithm` would undergo the argument validity check without contributing to the data tested by this script. I'll change the code to insist on the function being known: better fail with an exception than silently not test something that we might reasonably believe is tested.

gilles-peskine-arm replied:
I've changed the code to insist on knowing the function name. This uncovered a lack of coverage for some expressions involving a wildcard.

Patater commented on tests/scripts/test_psa_constant_names.py:
```diff
@@ -200,26 +229,35 @@ def parse_header(self, filename):
+        #pylint: disable=unused-argument
+        undeclared = list(self.generate_undeclared_names(argument))
+        if undeclared:
+            raise Exception('Undeclared names in test case', undeclared)
```
Should we return `False` instead of throwing an exception?

gilles-peskine-arm replied:
The exception here is intended. This only comes up if there's a test case with data that is not declared in a header. If this ever happens, it probably indicates that the test script need to be updated to add a missing header, or that there's a mistake in the test script logic, or that the intent of the test suite has changed in a way that doesn't agree with the test script. Whatever the reason, it's best that we know that something weird is going on, hence the test failure.
